src/main/scala/util/Hierarchy.scala

```diff
@@ -0,0 +1,110 @@
+// See LICENSE for license details.
+
+package sifive.blocks.util
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config._
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree.LogicalTreeNode
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.prci._
+import freechips.rocketchip.util.LocationMap
+
+import firrtl.graph._
+
+case object HierarchyKey extends Field[Option[DiGraph[HierarchicalLocation]]](None)
+
+case class DevicesSubsystemParams(
+  name: String,
+  logicalTreeNode: LogicalTreeNode,
+  asyncClockGroupsNode: ClockGroupEphemeralNode)
+
+class DevicesSubsystem(val location: HierarchicalLocation, val ibus: InterruptBusWrapper, params: DevicesSubsystemParams)(implicit p: Parameters)
+  extends LazyModule 
+    with Attachable
+    with HasConfigurableTLNetworkTopology 
+    with CanHaveDevices {
+
+  def devicesSubhierarchies = None
+  def logicalTreeNode = params.logicalTreeNode
+  implicit val asyncClockGroupsNode = params.asyncClockGroupsNode
+
+  lazy val module = new LazyModuleImp(this) {
+    override def desiredName: String = params.name
+  }
+}
+
+trait HasConfigurableHierarchy { this: Attachable =>
+  def location: HierarchicalLocation
+
+  def createHierarchyMap(
+    root: HierarchicalLocation,
+    graph: DiGraph[HierarchicalLocation],
+    context: Attachable): Unit = {
+
+    hierarchyMap += (root -> context)
+
+    // Create and recurse on child hierarchies
+    val edges = graph.getEdges(root)
+    edges.foreach { edge =>
+      val dssParams = DevicesSubsystemParams(
+        name = edge.name,
+        logicalTreeNode = this.logicalTreeNode,
```

Comment:
why are these `this` and not `context`? seems like probably each of these added subsystems should have their own OM and clock Nodes? (though at that point they are v close to being BaseSubsystems...)

Reply:
Even if they don't have their own nodes, it seems like you could still use `context`, which was populated with a reference to it's parent's nodes at the last recursive step

Reply:
My intent is that these members are copied from the root of the hierarchy tree, so `this`/`context` have the same meaning in this case

Reply:
But why do you want them copied from the root? In a tree with multiple levels, should they be copied from the parent? (which might or might not have itself copied them from the root depending on how we want to managing clocking / represent these things in the OM?)

Reply:
Yeah, that might make sense, though I'm not sure which approach is more idiomatic with how the OM (or asyncClockGroupsNode) is used today, though